src/html_json.js

```diff
@@ -12,61 +12,156 @@
+ *
+ * @param {object} params
+ * @returns configuration
+ */
+async function loadConfigFromFile(params) {
+  const { path, logger } = params;
+
+  const configfile = resolve(dirname(__dirname), dirname(path).substr(1), 'helix-index.yaml');
+  logger.debug(`Reading index configuration from: ${configfile}`);
+
+  try {
+    const source = await fs.readFile(configfile, 'utf8');
+    const document = YAML.parseDocument(source);
+    return document.toJSON() || {};
+  } catch (e) {
+    logger.error(`Failed to load index configuration from: ${configfile}`, e);
+    // config not usable return empty
+    return {
+      indices: [],
+    };
+  }
+}
+
+/**
+ * Fetch HTML file from the file system.
  *
- * @param {string} owner
- * @param {string} repo
- * @param {string} ref
+ * @param {object} params
+ * @return HTML file
+ */
+async function fetchHTMLFromFile(params) {
+  const { path, logger } = params;
+  const htmlfile = resolve(dirname(__dirname), path.replace(/\.md$/, '.html').substr(1));
+  logger.debug(`Reading HTML from: ${htmlfile}`);
+
+  return fs.readFile(htmlfile, 'utf8');
+}
+
+/**
+ * Index loader and HTML fetcher that operates on the local file system.
+ */
+const fileLoader = {
+  loadConfig: loadConfigFromFile,
+  fetchHTML: fetchHTMLFromFile,
+};
+
+/**
+ * Load an index configuration from a git repo.
  *
+ * @param {object} params
  * @returns configuration
  */
-async function loadConfig(owner, repo, ref) {
+async function loadConfigFromRepo(params) {
+  const {
+    owner, repo, ref, logger,
+  } = params;
+
   const url = `https://raw.githubusercontent.com/${owner}/${repo}/${ref}/helix-index.yaml`;
+  logger.debug(`Reading index configuration from: ${url}`);
+
   try {
     const response = await request(url);
     return YAML.parseDocument(response).toJSON() || {};
   } catch (e) {
+    logger.error(`Failed to load index configuration from: ${url}`, e);
     // config not usable return empty
     return {
       indices: [],
     };
   }
 }
 
+/**
+ * Fetch HTML page from a remote location, defined by owner, repo and path.
+ *
+ * @param {object} params
+ * @return HTML file
+ */
+async function fetchHTMLFromRepo(params) {
+  const {
+    index, owner, repo, path, logger,
+  } = params;
+  const pageUrl = index.fetch
+    .replace(/\{owner\}/g, owner)
+    .replace(/\{repo\}/g, repo)
+    .replace(/\{path\}/g, path);
+  logger.debug(`Reading HTML from: ${pageUrl}`);
+  return request(pageUrl);
+}
+
+/**
+ * Index loader and HTML fetcher that operates on git repositories.
+ */
+const repoLoader = {
+  loadConfig: loadConfigFromRepo,
+  fetchHTML: fetchHTMLFromRepo,
+};
+
 /**
  * Return a value in the DOM by evaluating an expression
  *
  * @param {HTMLElement} element
  * @param {string} expression
  */
 function getDOMValue(element, expression) {
-  const m = expression.match(/{([a-zA-Z]+)(\("([^"]+)"\))?}/);
+  const m = expression.match(/\${([a-zA-Z]+)\(('([^"]+)')?\)}/);
```

Comment:
Not sure how robust and generalizable this is, but I assume it's fine for theblog.